pgjdbc/src/test/java/org/postgresql/test/jdbc2/BlobTest.java

```diff
@@ -207,21 +210,94 @@ void markResetStream() throws Exception {
         LargeObjectManager lom = ((PGConnection) con).getLargeObjectAPI();
 
         long oid = rs.getLong(1);
-        LargeObject blob = lom.open(oid);
-        InputStream bis = blob.getInputStream();
-
-        assertEquals('<', bis.read());
-        bis.mark(4);
-        assertEquals('?', bis.read());
-        assertEquals('x', bis.read());
-        assertEquals('m', bis.read());
-        assertEquals('l', bis.read());
-        bis.reset();
-        assertEquals('?', bis.read());
+        try (LargeObject blob = lom.open(oid)) {
+          InputStream bis = blob.getInputStream();
+
+          assertEquals('<', bis.read());
+          bis.mark(4);
+          assertEquals('?', bis.read());
+          assertEquals('x', bis.read());
+          assertEquals('m', bis.read());
+          assertEquals('l', bis.read());
+          bis.reset();
+          assertEquals('?', bis.read());
+        }
+      }
+    }
+  }
+
+  @Test
+  void markResetWithInitialOffset() throws Exception {
+    assertTrue(uploadFile(TEST_FILE, NATIVE_STREAM) > 0);
+
+    try (Statement stmt = con.createStatement()) {
+      try (ResultSet rs = stmt.executeQuery("SELECT lo FROM testblob where id = '/test-file.xml'")) {
+        assertTrue(rs.next());
+
+        LargeObjectManager lom = ((PGConnection) con).getLargeObjectAPI();
+
+        long oid = rs.getLong(1);
+        try (LargeObject blob = lom.open(oid)) {
+          blob.seek(4);
+          InputStream bis = blob.getInputStream();
+
+          assertEquals('l', bis.read());
+          bis.reset();
+          assertEquals('l', bis.read());
+          assertEquals(' ', bis.read());
+          bis.mark(4);
+          assertEquals('v', bis.read());
+          assertEquals('e', bis.read());
+          bis.reset();
+          assertEquals('v', bis.read());
+        }
       }
     }
   }
 
+  @Test
+  void skip() throws Exception {
+    long expectedBigSkip;
+    if (TestUtil.haveMinimumServerVersion(con, ServerVersion.v9_3)) {
+      expectedBigSkip = 4398046442492L;
+    } else {
+      expectedBigSkip = 2147417083;
+    }
+
+    LargeObjectManager lom = ((PGConnection) con).getLargeObjectAPI();
+    long loid = createMediumLargeObject();
+
+    try (LargeObject blob = lom.open(loid, LargeObjectManager.READ)) {
+      InputStream bis = blob.getInputStream();
+      assertEquals(0, bis.read());
+      assertEquals(1024L, bis.skip(1024));
+      assertEquals(1, bis.read());
+      assertEquals(0L, bis.skip(-1));
+      assertEquals(1, bis.read());
+      assertEquals(64 * 1024L, bis.skip(64 * 1024));
+      assertEquals(65, bis.read());
+      assertEquals(expectedBigSkip, bis.skip(Long.MAX_VALUE));
+      assertEquals(-1, bis.read());
```

Comment:
Please add the corresponding messages describing the reason why expect certain values. If something goes wrong, it would be hard to tell why 65 was expected at line 278.

Reply:
I'm not really sure how to phrase it. What exactly do you want?

The reason why it should be 65 etc.  is obvious if you know the structure of the file. I can add a comment to each test describing it, if the code in `createMediumLargeObject` is insufficient?

Or do you want the aspect of the implementation that is being tested? Is there any preference on phrasing, given that it's supposed to describe the failure, not describe what it was testing?

No other assertions in this project use messages, so I have nothing to go on.

Reply:
> The reason why it should be 65 etc. is obvious if you know the structure of the 

I am afraid it is not obvious.

> No other assertions in this project use messages

I am afraid that is not quite right.
Please check here: https://github.com/pgjdbc/pgjdbc/blob/d12b625059310ce40548ff448f9485f5bd52b035/pgjdbc/src/test/java/org/postgresql/util/LazyCleanerTest.java#L63-L103
https://github.com/pgjdbc/pgjdbc/blob/d12b625059310ce40548ff448f9485f5bd52b035/pgjdbc/src/test/java/org/postgresql/test/jdbc2/BatchedInsertReWriteEnabledTest.java#L421-L433

> Is there any preference on phrasing, given that it's supposed to describe the failure, not describe what it was testing?

https://github.com/pgjdbc/pgjdbc/blob/master/CONTRIBUTING.md#test

> CONTRIBUTING.md#test: General rule: failing test should look like a good bug report

In other words, the message should describe the reason you expect something to equal to a something else.

Note: if the test class and test method gives enough information, it might be fine to omit the message.
However, in your case you put like ~10 assertions one after another, so it is important to make the failures distinct.

---

> No other assertions in this project use messages, so I have nothing to go on

There's maintenance cost. If the test fails sometime in the future, the one who observes the failure would have to pay the cost of analyzing the failure and deciphering the intentions behind the test. Of course the current code is not all perfect, however, it is important to keep the added code maintainable.

Reply:
It might be some of the assertions could be split into their own test methods. Then the failures would be easier to analyze and reproduce.

Reply:
> I am afraid it is not obvious.

I don't know how to write a good explanation because it really is obvious to me.

If you know a stream goes `[1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4]`, and you do `read(); skip(5); read()` then how it it not obvious that the results should be `1` and `3`?

I cannot put the full structure of the file into every assert message. I can put things like "failed to skip beyond internal buffer". That's what would be wrong with it. The value returned is arbitrary, but as previously discussed I chose this structure of values so that it's easy to see what the expected value should be, and minimise the chance of a fluke pass.

Reply:
> It might be some of the assertions could be split into their own test methods. Then the failures would be easier to analyze and reproduce.

I've done that a little, now that seeking too far throws and breaks the transaction. I was following the style of the existing tests here, which all have multiple assertions. Especially with streams you need multiple previous operations to set up the state you're testing. e.g. having had the buffer reallocate, being in a position where it might go backwards, etc.

I'll push the next updates without messages and more separate methods, and then see what you think.

Reply:
@vlsi it is done